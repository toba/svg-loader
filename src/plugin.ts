import SVGO from 'svgo';
import fs from 'fs';
import path from 'path';
import { Encoding } from '@toba/tools';
import { Compiler } from 'webpack';
import {
   default as HtmlWebpackPlugin,
   HtmlTagObject
} from 'html-webpack-plugin';

const name = 'html-webpack-inline-svg';

interface SvgPluginOptions {
   /**
    * SVG file paths relative to the Webpack context.
    */
   files: string[];
   svgo?: SVGO.Options;
}

const slugify = (name: string) => {
   if (name.includes('/')) {
      name = name.substr(name.lastIndexOf('/') + 1);
   }
   return name.replace('.svg', '').replace(/\s/, '');
};

/**
 * @see https://medium.com/webpack/webpack-4-migration-guide-for-plugins-loaders-20a79b927202
 * @see https://blog.johnnyreilly.com/2018/01/finding-webpack-4-use-map.html
 * @see https://github.com/DustinJackson/html-webpack-inline-source-plugin
 */
export class HtmlSvgPlugin {
   options: SvgPluginOptions;

   constructor(options?: SvgPluginOptions) {
      this.options = options === undefined ? { files: [] } : options;
   }

   apply(compiler: Compiler): void {
      /**
       * ID attribute will be updated to match each SVG filename.
       */
      const addID = { id: '' };
      /**
       * Plugin configurations.
       * @see https://github.com/svg/svgo#what-it-can-do
       */
      const plugins: { [key: string]: string | boolean | object } = {
         addAttributesToSVGElement: { attributes: [addID] },
         removeXMLNS: true,
         sortAttrs: true,
         removeViewBox: false,
         removeDimensions: true
      };
      const svgo = new SVGO({
         plugins: Object.keys(plugins).map(
            key => ({ [key]: plugins[key] } as any)
         )
      });

      compiler.hooks.compilation.tap(name, compilation => {
         if (this.options.files.length == 0) {
            return;
         }
         const options = compilation.compiler.options;
         const context: string =
            options.context !== undefined ? options.context : __dirname;

         HtmlWebpackPlugin.getHooks(compilation).alterAssetTagGroups.tapAsync(
            name,
            async (data, cb) => {
               const optimizers: Promise<any>[] = [];
               const files = this.options.files.reduce((map, name) => {
                  map.set(
                     slugify(name),
                     fs.readFileSync(path.resolve(context, name), {
                        encoding: Encoding.UTF8
                     })
                  );
                  return map;
               }, new Map<string, string>());

               files.forEach((text, id) => {
                  addID.id = id;
                  optimizers.push(svgo.optimize(text));
               });

               const svgs = await Promise.all(optimizers);

               const tag: HtmlTagObject = {
                  tagName: 'div',
                  attributes: {},
                  innerHTML: svgs.map(svg => svg.data).join(),
                  voidTag: false
               };

               data.bodyTags.push(tag);

               cb(null, data);
            }
         );
      });
   }
}
