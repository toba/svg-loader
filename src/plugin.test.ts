import '@toba/test';
import path from 'path';
import webpack from 'webpack';
import memoryfs from 'memory-fs';
import HtmlWebpackPlugin from 'html-webpack-plugin';
import { HtmlSvgPlugin } from './plugin';

const memFS = new memoryfs();

/**
 * Test loader.
 * @see https://webpack.js.org/contribute/writing-a-loader/#testing
 */
export function compile(): Promise<webpack.Stats> {
   const compiler = webpack({
      context: __dirname,
      mode: 'development',
      entry: `./__mocks__/entry.js`,
      output: {
         path: path.resolve(__dirname),
         filename: 'bundle.js'
      },
      plugins: [
         new HtmlWebpackPlugin(),
         new HtmlSvgPlugin({
            files: ['test.svg']
         })
      ]
   });

   compiler.outputFileSystem = memFS;

   return new Promise((resolve, reject) => {
      compiler.run((err, stats) => {
         if (err !== null) {
            reject(err);
         } else if (stats.hasErrors()) {
            reject(stats.compilation.errors[0]);
         } else {
            resolve(stats);
         }
      });
   });
}

test('works', async () => {
   const stats = await compile();
   expect(stats).toBeDefined();
   //const output = stats.toJson().modules[0].source;

   //expect(output).toMatchSnapshot();
});
