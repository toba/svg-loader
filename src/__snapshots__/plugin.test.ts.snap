// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset=\\"utf-8\\">
    <title>Webpack App</title>
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\"></head>
  <body>
  <script src=\\"bundle.js\\"></script><div><svg viewBox=\\"0 0 384 512\\" id=\\"air-freshener\\"><path d=\\"M378.94 321.41L284.7 224h49.22c15.3 0 23.66-16.6 13.86-27.53L234.45 69.96c3.43-6.61 5.55-14 5.55-21.96 0-26.51-21.49-48-48-48s-48 21.49-48 48c0 7.96 2.12 15.35 5.55 21.96L36.22 196.47C26.42 207.4 34.78 224 50.08 224H99.3L5.06 321.41C-6.69 333.56 3.34 352 21.7 352H160v32H48c-8.84 0-16 7.16-16 16v96c0 8.84 7.16 16 16 16h288c8.84 0 16-7.16 16-16v-96c0-8.84-7.16-16-16-16H224v-32h138.3c18.36 0 28.39-18.44 16.64-30.59zM192 31.98c8.85 0 16.02 7.17 16.02 16.02 0 8.84-7.17 16.02-16.02 16.02S175.98 56.84 175.98 48c0-8.85 7.17-16.02 16.02-16.02zM304 432v32H80v-32h224z\\"/></svg>,<svg viewBox=\\"0 0 24 24\\" id=\\"gauge-dashboard\\"><defs><style>.a{fill:none;stroke:#000;stroke-linecap:round;stroke-linejoin:round}</style></defs><circle cx=\\"12\\" cy=\\"14\\" r=\\"1.5\\" class=\\"a\\"/><path d=\\"M13.06 12.939l4.951-4.949M2.5 14.5h2M21.5 14.5h-2M20.776 10.365l-1.847.765M15.635 5.223l-.765 1.848M12 4.5v2M8.365 5.223l.765 1.848M5.283 7.282l1.412 1.415M3.224 10.365l1.846.765M12 2.5A11.5 11.5 0 0 0 .5 14v3.5a1 1 0 0 0 1 1h21a1 1 0 0 0 1-1V14A11.5 11.5 0 0 0 12 2.5z\\" class=\\"a\\"/></svg>,<svg class=\\"heroicon-award heroicon heroicons-lg\\" viewBox=\\"0 0 100 100\\" id=\\"heroicon-award-lg\\"><path fill=\\"#7ACFC0\\" d=\\"M30.42 61.96a6.07 6.07 0 0 0 1.54 1.95L26.18 85.5l9.8-5.66 5.65 9.8 5.65-21.07c.6.28 1.26.46 1.93.54l-6.9 25.7-7.06-12.24L23 89.64l7.42-27.68zM51.03 70l.27-.98c.5-.1.98-.26 1.43-.47h-.01l5.64 21.08 5.66-9.8 9.8 5.66-5.79-21.6a6.05 6.05 0 0 0 1.55-1.94L77 89.64l-12.25-7.07-7.07 12.25-6.65-24.8z\\" class=\\"heroicon-award-ribbon-outer heroicon-component-accent heroicon-component-fill\\"/><path fill=\\"#FFF\\" d=\\"M45.96 67.71l.16.14a5.31 5.31 0 0 0 .26.2l-5.09 19-4.95-8.58-8.57 4.95 5.06-18.9a6.18 6.18 0 0 0 4.1.87l.21-.04c2.73-.44 6.68.62 8.82 2.36zM58.7 87.05l-5.09-19 .27-.2.16-.14a12.72 12.72 0 0 1 8.82-2.36l.2.04c1.4.22 2.87-.13 4.1-.86l5.07 18.9-8.58-4.96-4.95 8.58z\\" class=\\"heroicon-award-ribbon-inner heroicon-component-fill\\"/><path fill=\\"#FFF\\" d=\\"M43.56 57.09a22.98 22.98 0 0 1-9.8-5.8l-.05-.05a22.98 22.98 0 0 1-5.8-9.8l-.26-.97a23.06 23.06 0 0 1 0-10.94l.26-.97a23.05 23.05 0 0 1 15.65-15.65l.96-.25a23.05 23.05 0 0 1 10.95 0l.96.25a23.05 23.05 0 0 1 15.65 15.65l.26.97a23.06 23.06 0 0 1 0 10.94l-.26.97a22.98 22.98 0 0 1-5.8 9.8l-.05.05a22.98 22.98 0 0 1-9.8 5.8l-.96.26a23.06 23.06 0 0 1-10.95 0l-.96-.26z\\" class=\\"heroicon-award-badge-inner heroicon-component-fill\\"/><path fill=\\"#7ACFC0\\" d=\\"M47.38 66.3l-.16-.14a13.2 13.2 0 0 0-4.45-2.24l1.5-5.6a24.06 24.06 0 0 0 11.46 0l1.5 5.6a13.2 13.2 0 0 0-4.45 2.24l-.17.13a4.42 4.42 0 0 1-5.23 0zm-4.08-8.25l-1.5 5.61a13.2 13.2 0 0 0-4.98-.28l-.2.03a4.42 4.42 0 0 1-4.53-2.62l-.08-.2a13.2 13.2 0 0 0-2.73-4.16l4.1-4.11a23.97 23.97 0 0 0 9.92 5.73zm13.4 0a23.97 23.97 0 0 0 9.91-5.73l4.1 4.1a13.2 13.2 0 0 0-2.73 4.18l-.07.2a4.42 4.42 0 0 1-4.53 2.6l-.2-.02a13.2 13.2 0 0 0-4.98.28l-1.5-5.6zm10.62-6.44a23.97 23.97 0 0 0 5.73-9.92l5.6 1.5a13.2 13.2 0 0 0-.28 4.98l.04.21a4.42 4.42 0 0 1-2.62 4.53l-.2.08a13.2 13.2 0 0 0-4.16 2.73l-4.11-4.1zm5.99-10.88a24.06 24.06 0 0 0 0-11.46l5.6-1.5a13.2 13.2 0 0 0 2.25 4.45l.13.16a4.42 4.42 0 0 1 0 5.24l-.13.16a13.2 13.2 0 0 0-2.24 4.45l-5.61-1.5zm-.26-12.43a23.97 23.97 0 0 0-5.73-9.91l4.1-4.11A13.2 13.2 0 0 0 75.6 17l.2.08c1.7.64 2.9 2.74 2.62 4.53l-.04.2a13.2 13.2 0 0 0 .29 4.98l-5.6 1.5zm-6.44-10.62a23.97 23.97 0 0 0-9.92-5.73l1.5-5.61c1.72.4 3.48.53 4.98.28l.2-.03a4.42 4.42 0 0 1 4.54 2.62l.07.2a13.2 13.2 0 0 0 2.74 4.16l-4.1 4.1zm-10.88-6a24.06 24.06 0 0 0-11.46 0l-1.5-5.6a13.2 13.2 0 0 0 4.45-2.24l.16-.13a4.42 4.42 0 0 1 5.23 0l.17.13a13.2 13.2 0 0 0 4.45 2.24l-1.5 5.6zm-12.43.27a23.97 23.97 0 0 0-9.91 5.73l-4.11-4.11A13.2 13.2 0 0 0 32 9.4l.08-.2a4.42 4.42 0 0 1 4.53-2.61l.2.03c1.5.25 3.27.12 4.98-.28l1.5 5.6zm-10.62 6.44a23.97 23.97 0 0 0-5.73 9.91l-5.61-1.5c.4-1.71.52-3.47.28-4.97l-.03-.21a4.42 4.42 0 0 1 2.62-4.53l.2-.08a13.2 13.2 0 0 0 4.16-2.73l4.1 4.1zm-6 10.88a24.06 24.06 0 0 0 0 11.46l-5.6 1.5a13.2 13.2 0 0 0-2.24-4.45l-.13-.16a4.42 4.42 0 0 1 0-5.24l.13-.16a13.2 13.2 0 0 0 2.24-4.45l5.6 1.5zm.27 12.42a23.98 23.98 0 0 0 5.73 9.92l-4.11 4.11A13.2 13.2 0 0 0 24.4 53l-.2-.08a4.42 4.42 0 0 1-2.61-4.53l.03-.2c.24-1.5.12-3.27-.28-4.98l5.6-1.5z\\" class=\\"heroicon-award-badge-outer heroicon-component-accent heroicon-component-fill\\"/><path fill=\\"#4A4A4A\\" d=\\"M70.94 59.31l9.24 34.48-14.7-8.49L57 100l-7-26.12L43 100l-8.5-14.7-14.7 8.49 9.25-34.48a12.78 12.78 0 0 0-5.37-4.45 6.46 6.46 0 0 1-4.07-6.8c.47-2.93-.59-6.88-2.33-9.02a6.46 6.46 0 0 1-.13-7.92c1.87-2.3 2.93-6.25 2.5-8.98a6.46 6.46 0 0 1 3.84-6.92c2.77-1.06 5.66-3.95 6.64-6.53a6.46 6.46 0 0 1 6.8-4.07l.2.03c2.73.44 6.68-.62 8.82-2.36a6.46 6.46 0 0 1 7.92-.13c2.3 1.87 6.25 2.93 8.98 2.5l.2-.04a6.41 6.41 0 0 1 6.72 3.88c1.06 2.77 3.95 5.66 6.53 6.65a6.46 6.46 0 0 1 4.07 6.79c-.47 2.93.59 6.88 2.33 9.02a6.46 6.46 0 0 1 .13 7.92c-1.87 2.3-2.93 6.25-2.5 8.98a6.46 6.46 0 0 1-3.84 6.92 12.9 12.9 0 0 0-5.56 4.53zm-24.98 8.4a12.72 12.72 0 0 0-8.82-2.36l-.2.04a6.18 6.18 0 0 1-4.1-.87l-5.07 18.9 8.57-4.95 4.95 8.58 5.1-19a3.97 3.97 0 0 0-.27-.2l-.16-.14zm-15.54-5.75L23 89.64l12.25-7.07 7.07 12.25L49.2 69.1a6.34 6.34 0 0 1-1.94-.55h.01l-5.65 21.08-5.65-9.8-9.8 5.66 5.78-21.6.1.08a6.07 6.07 0 0 1-1.64-2.02zM51.03 70l6.65 24.8 7.07-12.24L77 89.64l-7.42-27.68c-.36.74-.9 1.4-1.55 1.95l5.79 21.59-9.8-5.66-5.66 9.8-5.64-21.07c-.44.2-.93.36-1.42.46l-.27.98zm7.67 17.04l4.95-8.58 8.58 4.95-5.07-18.9a6.19 6.19 0 0 1-4.1.87l-.2-.04c-2.73-.44-6.68.62-8.82 2.36l-.16.14-.27.2 5.1 19zM47.38 66.29a4.42 4.42 0 0 0 5.23 0l.17-.13a13.2 13.2 0 0 1 4.45-2.24l-1.5-5.6a24.05 24.05 0 0 1-11.46 0l-1.5 5.6a13.2 13.2 0 0 1 4.45 2.24l.16.13zm-4.08-8.24a23.97 23.97 0 0 1-9.91-5.73l-4.11 4.1A13.2 13.2 0 0 1 32 60.6l.08.2a4.42 4.42 0 0 0 4.53 2.6l.2-.02a13.2 13.2 0 0 1 4.98.28l1.5-5.6zm.26-.96l.96.25a23.05 23.05 0 0 0 10.95 0l.96-.25a23.05 23.05 0 0 0 9.8-5.8l.06-.05c2.68-2.7 4.7-6.05 5.8-9.8l.25-.97a23.06 23.06 0 0 0 0-10.94l-.26-.97a23.05 23.05 0 0 0-15.65-15.65l-.96-.25a23.06 23.06 0 0 0-10.95 0l-.96.25a23.05 23.05 0 0 0-15.65 15.65l-.26.97a23.06 23.06 0 0 0 0 10.94l.26.97a22.98 22.98 0 0 0 5.8 9.8l.05.04a22.98 22.98 0 0 0 9.8 5.8zm13.13.96l1.5 5.61a13.2 13.2 0 0 1 4.98-.28l.2.03a4.42 4.42 0 0 0 4.54-2.62l.07-.2a13.2 13.2 0 0 1 2.74-4.16l-4.1-4.11a23.97 23.97 0 0 1-9.93 5.73zm10.63-6.44l4.1 4.11A13.2 13.2 0 0 1 75.6 53l.2-.08a4.42 4.42 0 0 0 2.62-4.53l-.04-.2a13.2 13.2 0 0 1 .29-4.98l-5.6-1.5a23.97 23.97 0 0 1-5.74 9.91zm5.99-10.88l5.6 1.5a13.2 13.2 0 0 1 2.25-4.45l.13-.16a4.42 4.42 0 0 0 0-5.24l-.13-.16a13.2 13.2 0 0 1-2.24-4.45l-5.61 1.5a24.06 24.06 0 0 1 0 11.46zm-.26-12.43l5.6-1.5a13.2 13.2 0 0 1-.28-4.97l.04-.21a4.42 4.42 0 0 0-2.62-4.53l-.2-.08a13.2 13.2 0 0 1-4.16-2.73l-4.11 4.1a23.97 23.97 0 0 1 5.73 9.92zm-6.44-10.62l4.1-4.11a13.2 13.2 0 0 1-2.73-4.17l-.07-.2a4.42 4.42 0 0 0-4.53-2.61l-.2.03c-1.51.25-3.27.12-4.98-.28l-1.5 5.6a23.97 23.97 0 0 1 9.91 5.74zm-10.88-6l1.5-5.6a13.2 13.2 0 0 1-4.45-2.24l-.17-.13a4.42 4.42 0 0 0-5.23 0l-.16.13a13.2 13.2 0 0 1-4.45 2.24l1.5 5.6a24.06 24.06 0 0 1 11.46 0zm-12.43.27l-1.5-5.61c-1.71.4-3.47.53-4.98.28l-.2-.03a4.42 4.42 0 0 0-4.53 2.62l-.08.2a13.2 13.2 0 0 1-2.73 4.16l4.1 4.1a23.97 23.97 0 0 1 9.92-5.72zm-10.62 6.44l-4.11-4.11A13.2 13.2 0 0 1 24.4 17l-.2.08a4.42 4.42 0 0 0-2.61 4.53l.03.2c.24 1.5.12 3.27-.28 4.98l5.6 1.5a23.97 23.97 0 0 1 5.74-9.91zm-6 10.88l-5.6-1.5a13.2 13.2 0 0 1-2.24 4.45l-.13.16a4.42 4.42 0 0 0 0 5.24l.13.16a13.2 13.2 0 0 1 2.24 4.45l5.6-1.5a24.06 24.06 0 0 1 0-11.46zm.27 12.42l-5.61 1.5c.4 1.72.52 3.48.28 4.98l-.03.21a4.42 4.42 0 0 0 2.62 4.53l.2.08a13.2 13.2 0 0 1 4.16 2.73l4.1-4.1a23.98 23.98 0 0 1-5.72-9.93zm14.1-24.58C43.75 15.76 46.78 15 50 15v1c-3.06 0-5.94.72-8.5 2l-.45-.9zM38 19l.6.8a19.1 19.1 0 0 0-2.04 1.76l-.7-.7c.67-.67 1.38-1.3 2.14-1.86zm-8 16c0-4.5 1.49-8.66 4-12l.8.6A18.92 18.92 0 0 0 31 35c0 3.05.72 5.94 2 8.5l-.9.45A19.92 19.92 0 0 1 30 35z\\" class=\\"heroicon-outline\\"/></svg>,<svg viewBox=\\"0 0 48 48\\" id=\\"ic-router-48px\\"><path fill=\\"none\\" d=\\"M0 0h48v48H0z\\"/><path d=\\"M40.33 11.81L42 10.13C39.24 7.38 35.62 6 32 6s-7.24 1.39-10 4.13l1.66 1.66C25.97 9.5 28.98 8.35 32 8.35s6.03 1.16 8.33 3.46zm-1.66 1.52c-1.84-1.84-4.25-2.76-6.67-2.76s-4.83.92-6.67 2.76l1.66 1.66A7.104 7.104 0 0 1 32 12.93c1.81 0 3.62.69 4.99 2.07l1.68-1.67zM38 26h-4v-8h-4v8H10c-2.21 0-4 1.79-4 4v8c0 2.21 1.79 4 4 4h28c2.21 0 4-1.79 4-4v-8c0-2.21-1.79-4-4-4zM16 36h-4v-4h4v4zm7 0h-4v-4h4v4zm7 0h-4v-4h4v4z\\"/></svg>,<svg xmlns:bx=\\"https://boxy-svg.com\\" class=\\"frog\\" viewBox=\\"0 0 100 100\\" id=\\"logo-colored\\"><path fill=\\"#4676b4\\" stroke=\\"#000\\" d=\\"M90.023 73.561c-2.469 3.515-5.145 6.645-7.834 9.258a41.48 41.48 0 0 1 1.269.005c5.563.111 10.101 1.333 10.134 2.73.022.895-1.811 1.645-4.593 2.039 4.591 1.245 7.704 2.856 7.443 4.025-.273 1.229-4.192 1.588-9.326.991 3.003 1.561 4.766 3.005 4.421 3.821-.58 1.37-6.871.458-14.053-2.036-3.115-1.082-5.886-2.284-7.977-3.41-1.905.452-3.569.344-4.864-.413-5.26-3.072-2.424-15.669 6.334-28.136 8.757-12.467 20.121-20.082 25.381-17.01 5.259 3.072 2.423 15.669-6.335 28.136zM6.24 85.554c.034-1.397 4.626-2.619 10.256-2.729a40.3 40.3 0 0 1 1.096-.007c-2.685-2.613-5.351-5.742-7.802-9.257C1.094 61.095-1.606 48.498 3.756 45.426c5.364-3.073 16.76 4.543 25.455 17.01 8.695 12.466 11.396 25.063 6.033 28.135-1.273.729-2.885.857-4.72.46-2.107 1.113-4.876 2.296-7.984 3.363C15.272 96.888 8.907 97.8 8.32 96.43c-.349-.816 1.435-2.26 4.474-3.821-5.196.597-9.161.238-9.437-.991-.264-1.169 2.886-2.78 7.531-4.025-2.816-.394-4.669-1.144-4.648-2.039z\\" class=\\"back-legs\\"/><ellipse cx=\\"50\\" cy=\\"65.729\\" fill=\\"#72a3e2\\" stroke=\\"#000\\" class=\\"torso\\" rx=\\"29.51\\" ry=\\"25.521\\"/><ellipse cx=\\"50\\" cy=\\"62.176\\" fill=\\"#b4cff2\\" stroke=\\"#000\\" class=\\"front-legs\\" rx=\\"20.093\\" ry=\\"22.849\\"/><path fill=\\"#b4cff2\\" stroke=\\"#000\\" d=\\"M73.576 98.11c-.086.84-2.428 1.256-5.231.929a14.85 14.85 0 0 1-2.744-.567c-.449.707-1.26 1.184-2.122 1.168-.924-.017-1.637-.597-1.842-1.401-1.096.445-2.556.854-4.178 1.138-3.564.622-6.564.406-6.7-.485-.116-.755 1.867-1.753 4.648-2.419-2.155-.741-3.584-1.708-3.445-2.45.185-.989 3.081-1.205 6.469-.483.333.071.659.149.975.232-.441-.869-.516-1.685-.117-2.126.55-.607 1.818-.298 2.99.672 3.544-6.309 10.393-18.793 10.171-20.656-.302-2.529-15.543-9.995-20.245-7.928 0 0 13.531-16.014 14.204-14.951 4.276 6.758 9.138 12.788 14.921 18.085 3.647 3.341-9.8 24.001-13.195 29.08.171.015.344.033.519.053 2.804.325 5.007 1.27 4.922 2.109zM47.795 63.743c-4.702-2.068-19.943 5.403-20.245 7.934-.222 1.861 6.611 14.325 10.159 20.646 1.174-.978 2.449-1.292 3.001-.683.399.441.325 1.257-.117 2.127.317-.083.642-.161.976-.232 3.388-.722 6.284-.506 6.469.483.139.743-1.29 1.711-3.444 2.453 2.78.666 4.762 1.665 4.647 2.421-.136.891-3.136 1.108-6.7.484-1.622-.283-3.082-.692-4.178-1.137-.205.804-.918 1.384-1.842 1.401-.861.016-1.673-.463-2.122-1.17-.781.249-1.723.45-2.744.569-2.803.327-5.145-.091-5.231-.93-.085-.84 2.118-1.785 4.922-2.112.168-.019.334-.036.498-.05-3.429-5.135-16.814-25.732-13.174-29.068 5.783-5.3 10.645-11.333 14.921-18.096.673-1.063 14.204 14.96 14.204 14.96z\\" class=\\"front-legs\\"/><path fill=\\"#b4cff2\\" d=\\"M33.941 49.488h32.01L49.946 81.919z\\" class=\\"front-legs\\"/><ellipse cx=\\"50\\" cy=\\"26.946\\" fill=\\"#72a3e2\\" stroke=\\"#000\\" class=\\"head\\" rx=\\"40\\" ry=\\"25\\"/><ellipse cx=\\"25\\" cy=\\"16\\" fill=\\"#72a3e2\\" stroke=\\"#000\\" class=\\"head\\" rx=\\"15\\" ry=\\"15\\"/><ellipse cx=\\"15.526\\" cy=\\"27.05\\" fill=\\"#4676b4\\" class=\\"cheek\\" rx=\\"3.707\\" ry=\\"6.564\\"/><ellipse cx=\\"84.475\\" cy=\\"27.05\\" fill=\\"#4676b4\\" class=\\"cheek\\" rx=\\"3.707\\" ry=\\"6.564\\"/><ellipse cx=\\"75\\" cy=\\"16\\" fill=\\"#72a3e2\\" stroke=\\"#000\\" class=\\"head\\" rx=\\"15\\" ry=\\"15\\"/><path fill=\\"#72a3e2\\" d=\\"M21.105 11.948l58.656.363v19.653L21.105 31.6z\\" class=\\"head\\"/><ellipse cx=\\"22.008\\" cy=\\"16\\" fill=\\"#fff\\" stroke=\\"#000\\" class=\\"eyes\\" rx=\\"15\\" ry=\\"15\\"/><ellipse cx=\\"78.326\\" cy=\\"16\\" fill=\\"#fff\\" stroke=\\"#000\\" class=\\"eyes\\" rx=\\"15\\" ry=\\"15\\"/><path d=\\"M71.654 14.914l3.198-.689a4.006 4.006 0 0 1 2.954-2.634 4.008 4.008 0 0 1 3.778 1.186l3.197-.689a2.054 2.054 0 1 1 .865 4.017l-3.197.688a4.008 4.008 0 0 1-6.733 1.449l-3.198.689a2.055 2.055 0 1 1-.864-4.017zM29.027 14.914a2.054 2.054 0 1 1-.864 4.017l-3.198-.689a4.008 4.008 0 0 1-6.733-1.449l-3.197-.688a2.054 2.054 0 1 1 .865-4.017l3.197.689a4.008 4.008 0 0 1 3.778-1.186 4.006 4.006 0 0 1 2.954 2.634z\\" class=\\"pupil\\"/><ellipse cx=\\"71.09\\" cy=\\"36.577\\" class=\\"nostril\\" rx=\\"1.77\\" ry=\\"1.77\\" transform=\\"matrix(1 0 0 1 -24.188 -14.047)\\"/><ellipse cx=\\"71.09\\" cy=\\"36.577\\" class=\\"nostril\\" rx=\\"1.77\\" ry=\\"1.77\\" transform=\\"matrix(1 0 0 1 -17.994 -14.047)\\"/><path fill=\\"#f90\\" stroke=\\"#000\\" d=\\"M39.951 6.139h20.022L49.962 17.185z\\" class=\\"forehead\\"/><path fill=\\"#b4cff2\\" stroke=\\"#000\\" d=\\"M13.413 37.067c8.934-2.855 22.014-4.655 36.587-4.655 14.573 0 27.653 1.8 36.587 4.655C80.37 45.832 66.327 51.946 50 51.946s-30.37-6.114-36.587-14.879z\\" class=\\"chin\\"/><path d=\\"M69.952 33.61c-4.017 1.645-11.17 2.739-19.325 2.739-8.351 0-15.651-1.147-19.608-2.858 5.834-.695 12.248-1.079 18.981-1.079 7.109 0 13.862.428 19.952 1.198z\\" class=\\"mouth\\"/><path fill=\\"#d8d8d8\\" stroke=\\"#000\\" d=\\"M81.714 67.143l6.229-8.646\\"/><path fill=\\"#d8d8d8\\" stroke=\\"#000\\" d=\\"M18.34 67.446L12.111 58.8\\" bx:origin=\\"0.381 0.382\\"/><path fill=\\"#d8d8d8\\" stroke=\\"#000\\" d=\\"M90.266 87.875l-5.578-1.023M87.616 92.764l-6.694-2.231M11.586 93.009l6.694-2.231M9.617 87.844l5.578-1.023\\"/><path stroke=\\"#000\\" d=\\"M57.629 96.05l-3.02.542M41.542 95.933l3.02.542\\"/><path fill=\\"none\\" stroke=\\"#000\\" d=\\"M4.497 57.087s-1.302-2.435-1.085-5.499c.203-2.876 1.24-3.331 1.24-3.331M81.33 54.841s1.549-1.639 3.331-3.099c1.787-1.464 3.563-2.246 3.563-2.246\\"/></svg></div></body>
</html>"
`;
